### ISSUE TITLE
Passing generic types to functions

### ISSUE THREAD

**Issue by jafioti** [PyO3/pyo3#1480]
I have a function written in Rust accepting a Vec<T>, and I am trying to build a wrapper around it in PyO3. Currently this is what my function header looks like:

`pub fn pad_mask<T: std::cmp::PartialEq>(batch: &Vec<Vec<T>>, pad_value: T) -> Vec<Vec<bool>> `

I have been having trouble implementing a wrapper for this. I get an error on T: cannot find type `T` in this scope
Is this possible to wrap, or will I need to use some hacks to get around it?

**Comment by nw0** [comment]
By "implementing a wrapper", do you mean with the `#[pyfunction]` macro, or something else?

I could be entirely wrong, but looking at the macro code, I think generic parameters aren't handled currently.

**Comment by jafioti** [comment]
Yes I mean #[pyfunction]. Is this something that is possible at all, is there any workarounds, or will this be a feature coming soon?

**Comment by davidhewitt** [comment]
Unfortunately this isn't possible at a technical level - a generic Rust function has a separate function created for each different type you use it with. And this set of types is known at compile time.

However we can only wrap _one_ function to Python.

Instead of a generic type parameter `T` you should probably use `&PyAny`.

**Comment by jafioti** [comment]
Ok, I've changed my function to:
`#[pyfunction]
pub fn pad_mask(batch: &Vec<Vec<&PyAny>>, pad_value: &PyAny) -> Vec<Vec<bool>> `
But I am now getting the error the trait bound &Vec<Vec<&pyo3::PyAny>>: PyClass is not satisfied

**Comment by davidhewitt** [comment]
Does `Vec<Vec<_>>` work without the reference?

FWIW I would recommend using `PyArray` from the numpy crate instead of `Vec<Vec<_>>` as it'll be much much faster than converting all the data into Rust. 

However you're going to need https://github.com/PyO3/rust-numpy/issues/175 to have a `PyObject` array. I could PR that for you sometime so it's on numpy `master` - or if you want it faster you could try opening a PR yourself?
